kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/builders/ExtensibleBotBuilder.kt

```diff
@@ -345,38 +345,53 @@ public open class ExtensibleBotBuilder {
 
     /** @suppress Internal function used to initially set up Koin. **/
     public open suspend fun setupKoin() {
-        startKoin {
-            var logLevel = koinLogLevel
+        startKoinIfNeeded()
 
-            if (logLevel == Level.INFO || logLevel == Level.DEBUG) {
-                // NOTE: Temporary workaround for Koin not supporting Kotlin 1.6
-                logLevel = Level.ERROR
-            }
+        hooksBuilder.runBeforeKoinSetup()
 
-            slf4jLogger(logLevel)
-//            environmentProperties()  // https://github.com/InsertKoinIO/koin/issues/1099
+        addBotKoinModule()
 
-            if (File("koin.properties").exists()) {
-                fileProperties("koin.properties")
-            }
+        hooksBuilder.runAfterKoinSetup()
+    }
 
-            modules()
+    /** @suppress Creates a new KoinApplication if it has not already been started. **/
+    private fun startKoinIfNeeded() {
+        var logLevel = koinLogLevel
+
+        if (logLevel == Level.INFO || logLevel == Level.DEBUG) {
+            // NOTE: Temporary workaround for Koin not supporting Kotlin 1.6
+            logLevel = Level.ERROR
         }
 
-        hooksBuilder.runBeforeKoinSetup()
+        if (koinNotStarted()) {
+            KordExContext.startKoin {
+                slf4jLogger(logLevel)
 
+                if (File("koin.properties").exists()) {
+                    fileProperties("koin.properties")
+                }
+            }
+        } else {
+            getKoin().logger.level = logLevel
+        }
+    }
+
+    /** @suppress Internal function that checks if Koin has been started. **/
+    private fun koinNotStarted(): Boolean = KordExContext.getOrNull() == null
+
+    /** @suppress Internal function that creates and loads the bot's main Koin modules.
+     * The modules provide important bot-related singletons. **/
+    private fun addBotKoinModule() {
         loadModule { single { this@ExtensibleBotBuilder } bind ExtensibleBotBuilder::class }
         loadModule { single { i18nBuilder.translationsProvider } bind TranslationsProvider::class }
         loadModule { single { chatCommandsBuilder.registryBuilder() } bind ChatCommandRegistry::class }
         loadModule { single { componentsBuilder.registryBuilder() } bind ComponentRegistry::class }
         loadModule { single { componentsBuilder.callbackRegistryBuilder() } bind ComponentCallbackRegistry::class }
-
```

Comment:
You've removed some newlines here - please don't do that, they're important for readability.